{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Pandas\n",
    "In this notebook, we'll encounter a very useful package for scientific computing in Python: Pandas. We can think of Pandas as \"numpy with labels\". This package is especially useful for data science and biology for a few reasons:\n",
    "* Great for real-world, heterogeneous data\n",
    "* Similar to Excel spreadsheets (but way faster!)\n",
    "* Smartly deals with missing data\n",
    "\n",
    "### At the end of this notebook, you'll be able to:\n",
    "* Create & manipulate Pandas dataframes\n",
    "* Load a tabular file of genetic data into a Pandas dataframe\n",
    "* Index and subset Pandas dataframes\n",
    "* Generate descriptive statistics for Pandas dataframes\n",
    "<hr>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Importing packages\n",
    "\n",
    "Before we can pandas, we need to import it. The convention is to import `pandas` as `pd`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "slideshow": {
     "slide_type": "fragment"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variable   Type      Data/Info\n",
      "------------------------------\n",
      "pd         module    <module 'pandas' from '/o<...>ages/pandas/__init__.py'>\n"
     ]
    }
   ],
   "source": [
    "# Import packages\n",
    "import pandas as pd\n",
    "\n",
    "# Use whos 'magic command' to see available modules\n",
    "%whos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Create and Manipulate Dataframes \n",
    "The two data structures of Pandas are the `Series` and the `DataFrame`. A `Series` is a one-dimensional object similar to a list. A `DataFrame` can be thought of as a two-dimensional numpy array or a collection of `Series` objects. Series and dataframes can contain multiple different data types such as integers, strings, and floats, similar to an Excel spreadsheet. Pandas also supports `string` lables unlike numpy arrays which only have numeric labels for their rows and columns. For a more in depth explanation, please visit the [Introduction to Data Structures](https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html) section in the Pandas User Guide. \n",
    "\n",
    "You can create a Pandas dataframe by inputting dictionaries into the Pandas function `pd.DataFrame()`, by reading files, or through functions built into the Pandas package. The function [`pd.read_csv()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) reads a comma- or tab-separated file and returns it as a `dataframe`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating a data from a dictionary\n",
    "\n",
    "Below, we'll create a dataframe from a dictionary. Since this is quite a bit to fit into one cell, we're also using a trick where we use a pair of open parentheses to tell Python to *ignore* the line breaks here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FIRST_NAME</th>\n",
       "      <th>LAST_NAME</th>\n",
       "      <th>YEAR</th>\n",
       "      <th>HOME_STATE</th>\n",
       "      <th>AGE</th>\n",
       "      <th>CALC_101_FINAL</th>\n",
       "      <th>ENGLISH_101_FINAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Daniel</td>\n",
       "      <td>Smith</td>\n",
       "      <td>1</td>\n",
       "      <td>NY</td>\n",
       "      <td>18</td>\n",
       "      <td>90.0</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ben</td>\n",
       "      <td>Leibstrom</td>\n",
       "      <td>1</td>\n",
       "      <td>NY</td>\n",
       "      <td>19</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kavita</td>\n",
       "      <td>Kanabar</td>\n",
       "      <td>1</td>\n",
       "      <td>PA</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linda</td>\n",
       "      <td>Thiel</td>\n",
       "      <td>4</td>\n",
       "      <td>CA</td>\n",
       "      <td>22</td>\n",
       "      <td>60.0</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Omar</td>\n",
       "      <td>Reichel</td>\n",
       "      <td>2</td>\n",
       "      <td>OK</td>\n",
       "      <td>21</td>\n",
       "      <td>70.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Jane</td>\n",
       "      <td>OConner</td>\n",
       "      <td>2</td>\n",
       "      <td>HI</td>\n",
       "      <td>19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Felicia</td>\n",
       "      <td>Rao</td>\n",
       "      <td>3</td>\n",
       "      <td>NY</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rachel</td>\n",
       "      <td>Crock</td>\n",
       "      <td>1</td>\n",
       "      <td>FL</td>\n",
       "      <td>17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Bob</td>\n",
       "      <td>McDonald</td>\n",
       "      <td>1</td>\n",
       "      <td>FL</td>\n",
       "      <td>18</td>\n",
       "      <td>98.0</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  FIRST_NAME  LAST_NAME  YEAR HOME_STATE  AGE  CALC_101_FINAL  \\\n",
       "0     Daniel      Smith     1         NY   18            90.0   \n",
       "1        Ben  Leibstrom     1         NY   19            80.0   \n",
       "2     Kavita    Kanabar     1         PA   19             NaN   \n",
       "3      Linda      Thiel     4         CA   22            60.0   \n",
       "4       Omar    Reichel     2         OK   21            70.0   \n",
       "5       Jane    OConner     2         HI   19             NaN   \n",
       "6    Felicia        Rao     3         NY   20             NaN   \n",
       "7     Rachel      Crock     1         FL   17             NaN   \n",
       "8        Bob   McDonald     1         FL   18            98.0   \n",
       "\n",
       "   ENGLISH_101_FINAL  \n",
       "0               80.0  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3               40.0  \n",
       "4               50.0  \n",
       "5                NaN  \n",
       "6                NaN  \n",
       "7               60.0  \n",
       "8               65.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create the dictionary\n",
    "students = {'FIRST_NAME': ['Daniel', 'Ben', 'Kavita', 'Linda',\n",
    "                  'Omar','Jane', 'Felicia', 'Rachel',\n",
    "                  'Bob'],\n",
    "   'LAST_NAME': ['Smith', 'Leibstrom', 'Kanabar', 'Thiel',\n",
    "                 'Reichel', 'OConner', 'Rao', 'Crock',\n",
    "                 'McDonald'],\n",
    "   'YEAR': [1, 1, 1, 4, 2, 2, 3, 1, 1],\n",
    "   'HOME_STATE': ['NY', 'NY', 'PA', 'CA', 'OK', 'HI',\n",
    "                  'NY','FL', 'FL'],\n",
    "   'AGE': [18, 19, 19, 22, 21, 19, 20, 17, 18],\n",
    "   'CALC_101_FINAL': [90, 80, None, 60, 70, None, None,\n",
    "                      None, 98],\n",
    "   'ENGLISH_101_FINAL': [80, None, None, 40, 50, None,\n",
    "                         None, 60, 65]} \n",
    "\n",
    "# Create the dataframe from our dictionary\n",
    "students_df = pd.DataFrame(students)\n",
    "students_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can access a list of all of the students' last names using the following notation: `df['column_name']`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        Smith\n",
       "1    Leibstrom\n",
       "2      Kanabar\n",
       "3        Thiel\n",
       "4      Reichel\n",
       "5      OConner\n",
       "6          Rao\n",
       "7        Crock\n",
       "8     McDonald\n",
       "Name: LAST_NAME, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Access LAST_NAME column\n",
    "lastnames = students_df['LAST_NAME']\n",
    "lastnames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One column of the dataframe is a **series**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check type of column\n",
    "type(lastnames)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading from a file example\n",
    "Below we will create a dataframe by reading the file `brainarea_vs_genes.tsv` which contains information on gene expression across multiple brain areas. We'll use the `read_csv` function to import our gene expression data as a DataFrame.\n",
    "\n",
    ">**About this dataset:**\n",
    "This dataset was created by Derek Howard and Abigail Mayes for the purpose of accelerating advances in data mining of open brain transcriptome data for polygenetic brain disorders. The data comes from normalized microarray datasets of gene expression from 6 adult human brains that was released by the Allen Brain Institute and then processed into the dataframe we will see below. For more information on this dataset please visit the <a href = \"https://github.com/derekhoward/HBAsets\"> HBAsets repository</a>. \n",
    "\n",
    "Below, we'll use the `read_csv()` function to import our gene expression data as a **data frame**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CNTNAP3</td>\n",
       "      <td>0.752163</td>\n",
       "      <td>1.548381</td>\n",
       "      <td>1.961218</td>\n",
       "      <td>1.511869</td>\n",
       "      <td>-1.052871</td>\n",
       "      <td>-1.044282</td>\n",
       "      <td>-0.951765</td>\n",
       "      <td>-1.066606</td>\n",
       "      <td>-1.379488</td>\n",
       "      <td>...</td>\n",
       "      <td>0.247441</td>\n",
       "      <td>0.522270</td>\n",
       "      <td>0.478694</td>\n",
       "      <td>-0.144929</td>\n",
       "      <td>0.019326</td>\n",
       "      <td>2.826586</td>\n",
       "      <td>-0.446810</td>\n",
       "      <td>-0.060737</td>\n",
       "      <td>-0.073722</td>\n",
       "      <td>-0.327626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AP1S1</td>\n",
       "      <td>0.983700</td>\n",
       "      <td>0.138025</td>\n",
       "      <td>0.240694</td>\n",
       "      <td>-0.138402</td>\n",
       "      <td>-0.859617</td>\n",
       "      <td>-0.853491</td>\n",
       "      <td>-0.799364</td>\n",
       "      <td>-0.830323</td>\n",
       "      <td>-0.688747</td>\n",
       "      <td>...</td>\n",
       "      <td>1.156757</td>\n",
       "      <td>1.320692</td>\n",
       "      <td>1.481280</td>\n",
       "      <td>0.885764</td>\n",
       "      <td>-0.466182</td>\n",
       "      <td>-0.612176</td>\n",
       "      <td>-1.022950</td>\n",
       "      <td>-1.208921</td>\n",
       "      <td>-1.157348</td>\n",
       "      <td>-0.976336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KLK3</td>\n",
       "      <td>0.340816</td>\n",
       "      <td>1.146565</td>\n",
       "      <td>1.446988</td>\n",
       "      <td>0.506936</td>\n",
       "      <td>0.596726</td>\n",
       "      <td>0.584251</td>\n",
       "      <td>0.525423</td>\n",
       "      <td>0.614704</td>\n",
       "      <td>1.453660</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.220420</td>\n",
       "      <td>-0.926273</td>\n",
       "      <td>-0.705074</td>\n",
       "      <td>-0.461776</td>\n",
       "      <td>0.896928</td>\n",
       "      <td>-0.502945</td>\n",
       "      <td>0.352190</td>\n",
       "      <td>0.290381</td>\n",
       "      <td>-0.313828</td>\n",
       "      <td>-0.597703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HUS1B</td>\n",
       "      <td>-0.647548</td>\n",
       "      <td>-0.828916</td>\n",
       "      <td>0.059452</td>\n",
       "      <td>-0.239782</td>\n",
       "      <td>1.148391</td>\n",
       "      <td>1.189058</td>\n",
       "      <td>1.233382</td>\n",
       "      <td>1.322978</td>\n",
       "      <td>0.658224</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.036261</td>\n",
       "      <td>-0.874315</td>\n",
       "      <td>-0.290853</td>\n",
       "      <td>-0.892696</td>\n",
       "      <td>0.549402</td>\n",
       "      <td>-1.890037</td>\n",
       "      <td>0.377516</td>\n",
       "      <td>-0.707721</td>\n",
       "      <td>0.006505</td>\n",
       "      <td>-0.074109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTF2E2</td>\n",
       "      <td>0.704505</td>\n",
       "      <td>0.382086</td>\n",
       "      <td>0.329439</td>\n",
       "      <td>0.301807</td>\n",
       "      <td>1.701373</td>\n",
       "      <td>1.559972</td>\n",
       "      <td>1.688480</td>\n",
       "      <td>1.648590</td>\n",
       "      <td>-1.226395</td>\n",
       "      <td>...</td>\n",
       "      <td>0.220697</td>\n",
       "      <td>0.391602</td>\n",
       "      <td>0.394778</td>\n",
       "      <td>-0.222447</td>\n",
       "      <td>-1.284692</td>\n",
       "      <td>0.369077</td>\n",
       "      <td>-0.428325</td>\n",
       "      <td>-0.015808</td>\n",
       "      <td>-1.065625</td>\n",
       "      <td>-0.165026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 233 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symbol  CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "0     CNTNAP3   0.752163   1.548381   1.961218   1.511869   \n",
       "1       AP1S1   0.983700   0.138025   0.240694  -0.138402   \n",
       "2        KLK3   0.340816   1.146565   1.446988   0.506936   \n",
       "3       HUS1B  -0.647548  -0.828916   0.059452  -0.239782   \n",
       "4      GTF2E2   0.704505   0.382086   0.329439   0.301807   \n",
       "\n",
       "   Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "0                   -1.052871           -1.044282   \n",
       "1                   -0.859617           -0.853491   \n",
       "2                    0.596726            0.584251   \n",
       "3                    1.148391            1.189058   \n",
       "4                    1.701373            1.559972   \n",
       "\n",
       "   Crus II, lateral hemisphere  Crus II, paravermis  Edinger-Westphal nucleus  \\\n",
       "0                    -0.951765            -1.066606                 -1.379488   \n",
       "1                    -0.799364            -0.830323                 -0.688747   \n",
       "2                     0.525423             0.614704                  1.453660   \n",
       "3                     1.233382             1.322978                  0.658224   \n",
       "4                     1.688480             1.648590                 -1.226395   \n",
       "\n",
       "   ...  temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "0  ...                        0.247441                      0.522270   \n",
       "1  ...                        1.156757                      1.320692   \n",
       "2  ...                       -1.220420                     -0.926273   \n",
       "3  ...                       -1.036261                     -0.874315   \n",
       "4  ...                        0.220697                      0.391602   \n",
       "\n",
       "   temporal pole, superior aspect  transverse gyri  trochlear nucleus  \\\n",
       "0                        0.478694        -0.144929           0.019326   \n",
       "1                        1.481280         0.885764          -0.466182   \n",
       "2                       -0.705074        -0.461776           0.896928   \n",
       "3                       -0.290853        -0.892696           0.549402   \n",
       "4                        0.394778        -0.222447          -1.284692   \n",
       "\n",
       "   tuberomammillary nucleus  ventral tegmental area  \\\n",
       "0                  2.826586               -0.446810   \n",
       "1                 -0.612176               -1.022950   \n",
       "2                 -0.502945                0.352190   \n",
       "3                 -1.890037                0.377516   \n",
       "4                  0.369077               -0.428325   \n",
       "\n",
       "   ventromedial hypothalamic nucleus  vestibular nuclei  zona incerta  \n",
       "0                          -0.060737          -0.073722     -0.327626  \n",
       "1                          -1.208921          -1.157348     -0.976336  \n",
       "2                           0.290381          -0.313828     -0.597703  \n",
       "3                          -0.707721           0.006505     -0.074109  \n",
       "4                          -0.015808          -1.065625     -0.165026  \n",
       "\n",
       "[5 rows x 233 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the list of lists as a data frame\n",
    "# Note \"delimiter\" from loadtxt is \"sep\" here\n",
    "gene_df = pd.read_csv('Data/brainarea_vs_genes.tsv',sep='\\t')\n",
    "\n",
    "#Show the first five rows\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the moment, the first column  of information above, called the **index**, just contains a list of numbers.\n",
    "\n",
    "> **Task**: Reassign the row labels by using the method `set_index()` on our dataframe. We can choose any column in our present dataframe to be the row values -- let's assign the row lables to be the `gene_symbol` and re-assign the dataframe. \n",
    "\n",
    "**Question for consideration**: Does `set_index()` work in place?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CA1 field</th>\n",
       "      <th>CA2 field</th>\n",
       "      <th>CA3 field</th>\n",
       "      <th>CA4 field</th>\n",
       "      <th>Crus I, lateral hemisphere</th>\n",
       "      <th>Crus I, paravermis</th>\n",
       "      <th>Crus II, lateral hemisphere</th>\n",
       "      <th>Crus II, paravermis</th>\n",
       "      <th>Edinger-Westphal nucleus</th>\n",
       "      <th>Heschl's gyrus</th>\n",
       "      <th>...</th>\n",
       "      <th>temporal pole, inferior aspect</th>\n",
       "      <th>temporal pole, medial aspect</th>\n",
       "      <th>temporal pole, superior aspect</th>\n",
       "      <th>transverse gyri</th>\n",
       "      <th>trochlear nucleus</th>\n",
       "      <th>tuberomammillary nucleus</th>\n",
       "      <th>ventral tegmental area</th>\n",
       "      <th>ventromedial hypothalamic nucleus</th>\n",
       "      <th>vestibular nuclei</th>\n",
       "      <th>zona incerta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CNTNAP3</th>\n",
       "      <td>0.752163</td>\n",
       "      <td>1.548381</td>\n",
       "      <td>1.961218</td>\n",
       "      <td>1.511869</td>\n",
       "      <td>-1.052871</td>\n",
       "      <td>-1.044282</td>\n",
       "      <td>-0.951765</td>\n",
       "      <td>-1.066606</td>\n",
       "      <td>-1.379488</td>\n",
       "      <td>0.069526</td>\n",
       "      <td>...</td>\n",
       "      <td>0.247441</td>\n",
       "      <td>0.522270</td>\n",
       "      <td>0.478694</td>\n",
       "      <td>-0.144929</td>\n",
       "      <td>0.019326</td>\n",
       "      <td>2.826586</td>\n",
       "      <td>-0.446810</td>\n",
       "      <td>-0.060737</td>\n",
       "      <td>-0.073722</td>\n",
       "      <td>-0.327626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AP1S1</th>\n",
       "      <td>0.983700</td>\n",
       "      <td>0.138025</td>\n",
       "      <td>0.240694</td>\n",
       "      <td>-0.138402</td>\n",
       "      <td>-0.859617</td>\n",
       "      <td>-0.853491</td>\n",
       "      <td>-0.799364</td>\n",
       "      <td>-0.830323</td>\n",
       "      <td>-0.688747</td>\n",
       "      <td>0.879118</td>\n",
       "      <td>...</td>\n",
       "      <td>1.156757</td>\n",
       "      <td>1.320692</td>\n",
       "      <td>1.481280</td>\n",
       "      <td>0.885764</td>\n",
       "      <td>-0.466182</td>\n",
       "      <td>-0.612176</td>\n",
       "      <td>-1.022950</td>\n",
       "      <td>-1.208921</td>\n",
       "      <td>-1.157348</td>\n",
       "      <td>-0.976336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLK3</th>\n",
       "      <td>0.340816</td>\n",
       "      <td>1.146565</td>\n",
       "      <td>1.446988</td>\n",
       "      <td>0.506936</td>\n",
       "      <td>0.596726</td>\n",
       "      <td>0.584251</td>\n",
       "      <td>0.525423</td>\n",
       "      <td>0.614704</td>\n",
       "      <td>1.453660</td>\n",
       "      <td>-0.295207</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.220420</td>\n",
       "      <td>-0.926273</td>\n",
       "      <td>-0.705074</td>\n",
       "      <td>-0.461776</td>\n",
       "      <td>0.896928</td>\n",
       "      <td>-0.502945</td>\n",
       "      <td>0.352190</td>\n",
       "      <td>0.290381</td>\n",
       "      <td>-0.313828</td>\n",
       "      <td>-0.597703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUS1B</th>\n",
       "      <td>-0.647548</td>\n",
       "      <td>-0.828916</td>\n",
       "      <td>0.059452</td>\n",
       "      <td>-0.239782</td>\n",
       "      <td>1.148391</td>\n",
       "      <td>1.189058</td>\n",
       "      <td>1.233382</td>\n",
       "      <td>1.322978</td>\n",
       "      <td>0.658224</td>\n",
       "      <td>-0.154011</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.036261</td>\n",
       "      <td>-0.874315</td>\n",
       "      <td>-0.290853</td>\n",
       "      <td>-0.892696</td>\n",
       "      <td>0.549402</td>\n",
       "      <td>-1.890037</td>\n",
       "      <td>0.377516</td>\n",
       "      <td>-0.707721</td>\n",
       "      <td>0.006505</td>\n",
       "      <td>-0.074109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTF2E2</th>\n",
       "      <td>0.704505</td>\n",
       "      <td>0.382086</td>\n",
       "      <td>0.329439</td>\n",
       "      <td>0.301807</td>\n",
       "      <td>1.701373</td>\n",
       "      <td>1.559972</td>\n",
       "      <td>1.688480</td>\n",
       "      <td>1.648590</td>\n",
       "      <td>-1.226395</td>\n",
       "      <td>-0.220292</td>\n",
       "      <td>...</td>\n",
       "      <td>0.220697</td>\n",
       "      <td>0.391602</td>\n",
       "      <td>0.394778</td>\n",
       "      <td>-0.222447</td>\n",
       "      <td>-1.284692</td>\n",
       "      <td>0.369077</td>\n",
       "      <td>-0.428325</td>\n",
       "      <td>-0.015808</td>\n",
       "      <td>-1.065625</td>\n",
       "      <td>-0.165026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 232 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             CA1 field  CA2 field  CA3 field  CA4 field  \\\n",
       "gene_symbol                                               \n",
       "CNTNAP3       0.752163   1.548381   1.961218   1.511869   \n",
       "AP1S1         0.983700   0.138025   0.240694  -0.138402   \n",
       "KLK3          0.340816   1.146565   1.446988   0.506936   \n",
       "HUS1B        -0.647548  -0.828916   0.059452  -0.239782   \n",
       "GTF2E2        0.704505   0.382086   0.329439   0.301807   \n",
       "\n",
       "             Crus I, lateral hemisphere  Crus I, paravermis  \\\n",
       "gene_symbol                                                   \n",
       "CNTNAP3                       -1.052871           -1.044282   \n",
       "AP1S1                         -0.859617           -0.853491   \n",
       "KLK3                           0.596726            0.584251   \n",
       "HUS1B                          1.148391            1.189058   \n",
       "GTF2E2                         1.701373            1.559972   \n",
       "\n",
       "             Crus II, lateral hemisphere  Crus II, paravermis  \\\n",
       "gene_symbol                                                     \n",
       "CNTNAP3                        -0.951765            -1.066606   \n",
       "AP1S1                          -0.799364            -0.830323   \n",
       "KLK3                            0.525423             0.614704   \n",
       "HUS1B                           1.233382             1.322978   \n",
       "GTF2E2                          1.688480             1.648590   \n",
       "\n",
       "             Edinger-Westphal nucleus  Heschl's gyrus  ...  \\\n",
       "gene_symbol                                            ...   \n",
       "CNTNAP3                     -1.379488        0.069526  ...   \n",
       "AP1S1                       -0.688747        0.879118  ...   \n",
       "KLK3                         1.453660       -0.295207  ...   \n",
       "HUS1B                        0.658224       -0.154011  ...   \n",
       "GTF2E2                      -1.226395       -0.220292  ...   \n",
       "\n",
       "             temporal pole, inferior aspect  temporal pole, medial aspect  \\\n",
       "gene_symbol                                                                 \n",
       "CNTNAP3                            0.247441                      0.522270   \n",
       "AP1S1                              1.156757                      1.320692   \n",
       "KLK3                              -1.220420                     -0.926273   \n",
       "HUS1B                             -1.036261                     -0.874315   \n",
       "GTF2E2                             0.220697                      0.391602   \n",
       "\n",
       "             temporal pole, superior aspect  transverse gyri  \\\n",
       "gene_symbol                                                    \n",
       "CNTNAP3                            0.478694        -0.144929   \n",
       "AP1S1                              1.481280         0.885764   \n",
       "KLK3                              -0.705074        -0.461776   \n",
       "HUS1B                             -0.290853        -0.892696   \n",
       "GTF2E2                             0.394778        -0.222447   \n",
       "\n",
       "             trochlear nucleus  tuberomammillary nucleus  \\\n",
       "gene_symbol                                                \n",
       "CNTNAP3               0.019326                  2.826586   \n",
       "AP1S1                -0.466182                 -0.612176   \n",
       "KLK3                  0.896928                 -0.502945   \n",
       "HUS1B                 0.549402                 -1.890037   \n",
       "GTF2E2               -1.284692                  0.369077   \n",
       "\n",
       "             ventral tegmental area  ventromedial hypothalamic nucleus  \\\n",
       "gene_symbol                                                              \n",
       "CNTNAP3                   -0.446810                          -0.060737   \n",
       "AP1S1                     -1.022950                          -1.208921   \n",
       "KLK3                       0.352190                           0.290381   \n",
       "HUS1B                      0.377516                          -0.707721   \n",
       "GTF2E2                    -0.428325                          -0.015808   \n",
       "\n",
       "             vestibular nuclei  zona incerta  \n",
       "gene_symbol                                   \n",
       "CNTNAP3              -0.073722     -0.327626  \n",
       "AP1S1                -1.157348     -0.976336  \n",
       "KLK3                 -0.313828     -0.597703  \n",
       "HUS1B                 0.006505     -0.074109  \n",
       "GTF2E2               -1.065625     -0.165026  \n",
       "\n",
       "[5 rows x 232 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_index = 'gene_symbol'\n",
    "gene_df = gene_df.set_index(row_index)\n",
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It would help to know what information is in our dataset. In other words, what is across the columns at the top?\n",
    "\n",
    "> **Task** Show a list of columns by accessing the `columns` attribute of our dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CA1 field',\n",
       " 'CA2 field',\n",
       " 'CA3 field',\n",
       " 'CA4 field',\n",
       " 'Crus I, lateral hemisphere',\n",
       " 'Crus I, paravermis',\n",
       " 'Crus II, lateral hemisphere',\n",
       " 'Crus II, paravermis',\n",
       " 'Edinger-Westphal nucleus',\n",
       " \"Heschl's gyrus\",\n",
       " 'I-II',\n",
       " 'III',\n",
       " 'III, lateral hemisphere',\n",
       " 'III, paravermis',\n",
       " 'IV',\n",
       " 'IV, lateral hemisphere',\n",
       " 'IV, paravermis',\n",
       " 'IX',\n",
       " 'IX, lateral hemisphere',\n",
       " 'IX, paravermis',\n",
       " 'V',\n",
       " 'V, lateral hemisphere',\n",
       " 'V, paravermis',\n",
       " 'VI',\n",
       " 'VI, lateral hemisphere',\n",
       " 'VI, paravermis',\n",
       " 'VIIAf',\n",
       " 'VIIAt',\n",
       " 'VIIB',\n",
       " 'VIIB, lateral hemisphere',\n",
       " 'VIIB, paravermis',\n",
       " 'VIIIA',\n",
       " 'VIIIA, lateral hemisphere',\n",
       " 'VIIIA, paravermis',\n",
       " 'VIIIB',\n",
       " 'VIIIB, lateral hemisphere',\n",
       " 'VIIIB, paravermis',\n",
       " 'X',\n",
       " 'X, lateral hemisphere',\n",
       " 'X, paravermis',\n",
       " 'abducens nucleus',\n",
       " 'amygdalohippocampal transition zone',\n",
       " 'angular gyrus, inferior bank of gyrus',\n",
       " 'angular gyrus, superior bank of gyrus',\n",
       " 'anterior group of nuclei',\n",
       " 'anterior hypothalamic area',\n",
       " 'anterior orbital gyrus',\n",
       " 'arcuate nucleus of medulla',\n",
       " 'arcuate nucleus of the hypothalamus',\n",
       " 'basal nucleus of meynert',\n",
       " 'basolateral nucleus',\n",
       " 'basomedial nucleus',\n",
       " 'bed  nucleus of stria terminalis',\n",
       " 'body of caudate nucleus',\n",
       " 'caudal group of intralaminar nuclei',\n",
       " 'central glial substance',\n",
       " 'central gray of the pons',\n",
       " 'central gray substance of midbrain',\n",
       " 'central medullary reticular group',\n",
       " 'central nucleus',\n",
       " 'choroid plexus of the lateral ventricle',\n",
       " 'cingulate gyrus, frontal part, inferior bank of gyrus',\n",
       " 'cingulate gyrus, frontal part, superior bank of gyrus',\n",
       " 'cingulate gyrus, parietal part, inferior bank of gyrus',\n",
       " 'cingulate gyrus, parietal part, superior bank of gyrus',\n",
       " 'cingulate gyrus, retrosplenial part, inferior bank of gyrus',\n",
       " 'cingulate gyrus, retrosplenial part, superior bank of gyrus',\n",
       " 'cingulum bundle',\n",
       " 'claustrum',\n",
       " 'cochlear nuclei',\n",
       " 'corpus callosum',\n",
       " 'cortico-medial group',\n",
       " 'cuneate nucleus',\n",
       " 'cuneiform nucleus',\n",
       " 'cuneus, peristriate',\n",
       " 'cuneus, striate',\n",
       " 'dentate gyrus',\n",
       " 'dentate nucleus',\n",
       " 'dorsal lateral geniculate nucleus',\n",
       " 'dorsal motor nucleus of the vagus',\n",
       " 'dorsomedial hypothalamic nucleus',\n",
       " 'emboliform nucleus',\n",
       " 'facial motor nucleus',\n",
       " 'fastigial nucleus',\n",
       " 'frontal operculum',\n",
       " 'frontal pole, inferior aspect',\n",
       " 'frontal pole, medial aspect',\n",
       " 'frontal pole, superior aspect',\n",
       " 'fusiform gyrus, bank of cos',\n",
       " 'fusiform gyrus, bank of the its',\n",
       " 'fusiform gyrus, lateral bank of gyrus',\n",
       " 'gigantocellular group',\n",
       " 'globose nucleus',\n",
       " 'globus pallidus, external segment',\n",
       " 'globus pallidus, internal segment',\n",
       " 'gracile nucleus',\n",
       " 'gyrus rectus',\n",
       " 'head of caudate nucleus',\n",
       " 'hypoglossal nucleus',\n",
       " 'inferior colliculus',\n",
       " 'inferior frontal gyrus, opercular part',\n",
       " 'inferior frontal gyrus, orbital part',\n",
       " 'inferior frontal gyrus, triangular part',\n",
       " 'inferior occipital gyrus, inferior bank of gyrus',\n",
       " 'inferior occipital gyrus, superior bank of gyrus',\n",
       " 'inferior olivary complex',\n",
       " 'inferior rostral gyrus',\n",
       " 'inferior temporal gyrus, bank of mts',\n",
       " 'inferior temporal gyrus, bank of the its',\n",
       " 'inferior temporal gyrus, lateral bank of gyrus',\n",
       " 'interstitial nucleus of Cajal',\n",
       " 'lateral group of nuclei, dorsal division',\n",
       " 'lateral group of nuclei, ventral division',\n",
       " 'lateral habenular nucleus',\n",
       " 'lateral hypothalamic area, anterior region',\n",
       " 'lateral hypothalamic area, mammillary region',\n",
       " 'lateral hypothalamic area, tuberal region',\n",
       " 'lateral mammillary nucleus',\n",
       " 'lateral medullary reticular group',\n",
       " 'lateral nucleus',\n",
       " 'lateral orbital gyrus',\n",
       " 'lateral parabrachial nucleus',\n",
       " 'lateral tuberal nucleus',\n",
       " 'lingual gyrus, peristriate',\n",
       " 'lingual gyrus, striate',\n",
       " 'locus ceruleus',\n",
       " 'long insular gyri',\n",
       " 'mammillary body',\n",
       " 'medial geniculate complex',\n",
       " 'medial group of nuclei',\n",
       " 'medial habenular nucleus',\n",
       " 'medial mammillary nucleus',\n",
       " 'medial orbital gyrus',\n",
       " 'medial parabrachial nucleus',\n",
       " 'midbrain raphe nuclei',\n",
       " 'midbrain reticular formation',\n",
       " 'middle frontal gyrus, inferior bank of gyrus',\n",
       " 'middle frontal gyrus, superior bank of gyrus',\n",
       " 'middle temporal gyrus, inferior bank of gyrus',\n",
       " 'middle temporal gyrus, superior bank of gyrus',\n",
       " 'motor nucleus of trigeminal nerve',\n",
       " 'nucleus accumbens',\n",
       " 'nucleus of Darkschewitsch',\n",
       " 'nucleus of the diagonal band, horizontal division',\n",
       " 'nucleus of the diagonal band, vertical division',\n",
       " 'nucleus subceruleus',\n",
       " 'occipital pole, inferior aspect',\n",
       " 'occipital pole, lateral aspect',\n",
       " 'occipital pole, superior aspect',\n",
       " 'occipito-temporal gyrus, inferior bank of gyrus',\n",
       " 'occipito-temporal gyrus, superior bank of gyrus',\n",
       " 'oculomotor nuclear complex',\n",
       " 'olfactory tubercle',\n",
       " 'pallidohypothalamic nucleus',\n",
       " 'paracentral lobule, anterior part',\n",
       " 'paracentral lobule, anterior part, inferior bank of gyrus',\n",
       " 'paracentral lobule, anterior part, superior bank of gyrus',\n",
       " 'paracentral lobule, posterior part, bank of cingulate sulcus',\n",
       " 'paracentral lobule, posterior part, lateral bank of gyrus',\n",
       " 'parahippocampal gyrus, bank of the cos',\n",
       " 'parahippocampal gyrus, lateral bank of gyrus',\n",
       " 'paraterminal gyrus',\n",
       " 'paraventricular nuclei, left of thalamus',\n",
       " 'paraventricular nuclei, right of thalamus',\n",
       " 'paraventricular nucleus of the hypothalamus',\n",
       " 'parolfactory gyri',\n",
       " 'perifornical nucleus',\n",
       " 'pineal gland',\n",
       " 'piriform cortex',\n",
       " 'planum polare',\n",
       " 'planum temporale',\n",
       " 'pontine nuclei',\n",
       " 'pontine raphe nucleus',\n",
       " 'pontine reticular formation',\n",
       " 'postcentral gyrus, bank of the central sulcus',\n",
       " 'postcentral gyrus, bank of the posterior central sulcus',\n",
       " 'postcentral gyrus, inferior lateral aspect of gyrus',\n",
       " 'postcentral gyrus, superior lateral aspect of gyrus',\n",
       " 'posterior group of nuclei',\n",
       " 'posterior hypothalamic area',\n",
       " 'posterior orbital gyrus',\n",
       " 'precentral gyrus, bank of the central sulcus',\n",
       " 'precentral gyrus, bank of the precentral sulcus',\n",
       " 'precentral gyrus, inferior lateral aspect of gyrus',\n",
       " 'precentral gyrus, superior lateral aspect of gyrus',\n",
       " 'precuneus, inferior lateral bank of gyrus',\n",
       " 'precuneus, superior lateral bank of gyrus',\n",
       " 'preoptic region',\n",
       " 'pretectal region',\n",
       " 'principal sensory nucleus of trigeminal nerve',\n",
       " 'putamen',\n",
       " 'raphe nuclei of medulla',\n",
       " 'red nucleus',\n",
       " 'reticular nucleus of thalamus',\n",
       " 'rostral group of intralaminar nuclei',\n",
       " 'septal nuclei',\n",
       " 'short insular gyri',\n",
       " 'spinal trigeminal nucleus',\n",
       " 'subcallosal cingulate gyrus',\n",
       " 'subcuneiform nucleus',\n",
       " 'subiculum',\n",
       " 'substantia innominata',\n",
       " 'substantia nigra, pars compacta',\n",
       " 'substantia nigra, pars reticulata',\n",
       " 'subthalamic nucleus',\n",
       " 'superior colliculus',\n",
       " 'superior frontal gyrus',\n",
       " 'superior frontal gyrus, lateral bank of gyrus',\n",
       " 'superior frontal gyrus, medial bank of gyrus',\n",
       " 'superior occipital gyrus, inferior bank of gyrus',\n",
       " 'superior occipital gyrus, superior bank of gyrus',\n",
       " 'superior olivary complex',\n",
       " 'superior parietal lobule, inferior bank of gyrus',\n",
       " 'superior parietal lobule, superior bank of gyrus',\n",
       " 'superior rostral gyrus',\n",
       " 'superior temporal gyrus, inferior bank of gyrus',\n",
       " 'superior temporal gyrus, lateral bank of gyrus',\n",
       " 'supramammillary nucleus',\n",
       " 'supramarginal gyrus, inferior bank of gyrus',\n",
       " 'supramarginal gyrus, superior bank of gyrus',\n",
       " 'supraoptic nucleus',\n",
       " 'tail of caudate nucleus',\n",
       " 'temporal pole, inferior aspect',\n",
       " 'temporal pole, medial aspect',\n",
       " 'temporal pole, superior aspect',\n",
       " 'transverse gyri',\n",
       " 'trochlear nucleus',\n",
       " 'tuberomammillary nucleus',\n",
       " 'ventral tegmental area',\n",
       " 'ventromedial hypothalamic nucleus',\n",
       " 'vestibular nuclei',\n",
       " 'zona incerta']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Access the columns of our dataframe\n",
    "gene_df.columns\n",
    "\n",
    "# Convert into a list to see all..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indexing Dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Indexing in Pandas works slightly different than in NumPy. Similar to a dictionary, we can index dataframes by their names. \n",
    "\n",
    "The syntax for indexing **single locations** in a dataframe is `dataframe.loc[row_label,column_label]`.\n",
    "\n",
    "To index an **individual column**, we use the shorthand syntax `dataframe[column_label]`.\n",
    "\n",
    "To index an **individual row**, we use the syntax `dataframe.loc[row_label]`.\n",
    "\n",
    "To index by **index #**, we use the syntax `dataframe.iloc[index_number]`.\n",
    "\n",
    "Below are some examples on how to access rows, columns, and single values in our dataframe. For more information on indexing dataframes, visit the <a href = \"https://pandas.pydata.org/docs/user_guide/indexing.html#indexing\"> \"Indexing and selecting data\"</a> section in the Pandas User Guide.\n",
    "\n",
    "> **Task**: Try each of these forms of indexing below. Can you use this syntax to get multiple values (a slice or subset) of the dataframe as well?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select a single location\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select a single column\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Select a single row\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select a single index\n",
    "# Hint: You can show all indices using gene_df.index\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To select multiple different columns, you can use a `list` of all your columns of interest as so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe w/ only CA field entries \n",
    "CA_field_df = gene_df[['CA1 field', \n",
    "                       'CA2 field', \n",
    "                       'CA3 field', \n",
    "                       'CA4 field']]\n",
    "CA_field_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Subsetting dataframes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Like NumPy arrays, we can subset our original dataframe to only include data that meets our criteria. Our dataframe has data on multiple different brain areas with many gene expression values. You can filter this dataframe using the following syntax:\n",
    "```\n",
    "new_df = original_df[original_df['Column of Interest'] == 'Desired Value']\n",
    "```\n",
    "In plain english, what this is saying is: save a dataframe from the original dataframe, where the original dataframe values in my Column of Interest are equal to my Desired Value. For more information on subsetting,  visit the <a href = \"https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html\"> \"How do I select a subset of a DataFrame\"</a> section in the Pandas documentation. \n",
    "\n",
    "Below we will demonstrate how to execute this by taking a look at the `CA1 field` column in `gene_df`. We will create a dataframe from `gene_df` that only contains genes that showed a certain level of gene expression. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dataframe with only genes that have an expression \n",
    "# value greater than 1.7 in 'CA1 field' \n",
    "\n",
    "high_CA1 = gene_df[gene_df['CA1 field'] > 1.7]\n",
    "high_CA1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DataFrame Methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pandas has many useful methods that you can use on your data, including `describe`, `mean`, and more. To learn more about all the different methods that can be used to manipulate and analyze dataframes, please visit the <a href = \"https://pandas.pydata.org/docs/user_guide/index.html\"> Pandas User Guide </a>. We will demonstrate some of these methods below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `describe` method returns descriptive statistics of all the columns in our dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `mean` and `std` method return the mean and standard deviation of each column in the dataframe, respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try df.mean()\n",
    "gene_df.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Try df.std()\n",
    "gene_df.std()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merging & joining dataframes\n",
    "\n",
    "Let's say we have two different dataframes and we would like to combine the two into one single dataframe. We can use either the `merge` or `join` Pandas methods in order to pull all of this data into one dataframe. \n",
    "\n",
    "![](http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png)\n",
    "\n",
    "There are different types of joins/merges you can do in Pandas, illustrated <a href=\"http://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/\">above</a>. Here, we want to do an **inner** merge, where we're only keeping entries with indices that are in both dataframes. We could do this merge based on columns, alternatively.\n",
    "\n",
    "**Inner** is the default kind of join, so we do not need to specify it. And by default, join will use the 'left' dataframe, in other words, the dataframe that is executing the `join` method.\n",
    "\n",
    "If you need more information, look at the <a href=\"https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html\">join</a> and <a href=\"https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html\">merge</a> documentation: you can use either of these to unite your dataframes, though join will be simpler!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below is an example of how to join two separate dataframe into one, unified dataframe. We start with one dataframe with only entries from the *temporal pole* and another dataframe with only entries from the CA fields of the hippocampus. We can then join the two dataframes together using the syntax `unified_df = df_1.join(df_2)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe w/ only Temporal Pole entries \n",
    "temporal_pole_df = gene_df[['temporal pole, inferior aspect', \n",
    "                            'temporal pole, medial aspect', \n",
    "                            'temporal pole, superior aspect']]\n",
    "temporal_pole_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join the two dataframes\n",
    "df_1 = temporal_pole_df\n",
    "df_2 = CA_field_df\n",
    "\n",
    "unified_df = df_1.join(df_2)\n",
    "unified_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "## Group Task (To submit on Canvas)\n",
    "### *Does the number of neurons in the cortex scale with cortex size?*\n",
    "In this final task, we'll look at [this dataset](http://www.suzanaherculanohouzel.com/2015-h-h-et-al-bbe-database/), which is already uploaded to your DataHub, to figure out whether the number of neurons in the cortex scales with cortex size, and if this scaling is different for primates.  \n",
    "\n",
    "To complete this task, do the following steps:\n",
    "\n",
    "1. Use `pd.read_csv()` to read in our comma-delimited file 'Data/species_brainmass_neurons.csv' and assign it to a dataframe called `brains_df`.\n",
    "2. Separate the dataframe into two dataframes: one that contains all of the primates (`primate_df`; 'Order' is Primata) and one that contains all of the data that are *not* from primates (`nonprimate_df`). **Hint**: You can subset the dataframe by using the syntax `brains_df[brains_df['Order']=='Primata']`.\n",
    "3. Using two separate, subsequent calls to `plt.scatter()` for primates and then non-primates, create a scatterplot with `cortex_mass` on the x axis and `neurons` on the y axis.\n",
    "4. Label your axes.\n",
    "5. (Optional) log scale your axes using `plt.xscale('log')` and `plt.yscale('log')`.\n",
    "6. Right click on the image or screenshot to save and submit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<hr>\n",
    "\n",
    "## Additional resources\n",
    "See the [Python Data Science Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/03.00-introduction-to-pandas.html) for a more in depth exploration of Pandas, and of course, the [Pandas documentation](https://pandas.pydata.org/docs/user_guide/index.html)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "rise": {
   "theme": "serif"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
